cores/esp32/IPAddress.cpp

```diff
@@ -20,62 +20,62 @@
 #include <Arduino.h>
 #include <IPAddress.h>
 #include <Print.h>
+#include <algorithm>
 
-IPAddress::IPAddress()
-{
-    _address.dword = 0;
-}
+// IPAddress::IPAddress()
+// {
+//     _address.dword = 0;
+// }
 
 IPAddress::IPAddress(uint8_t first_octet, uint8_t second_octet, uint8_t third_octet, uint8_t fourth_octet)
-{
-    _address.bytes[0] = first_octet;
-    _address.bytes[1] = second_octet;
-    _address.bytes[2] = third_octet;
-    _address.bytes[3] = fourth_octet;
-}
+	: _address({first_octet,
+				second_octet,
+    			third_octet,
+    			fourth_octet})
+{}
 
 IPAddress::IPAddress(uint32_t address)
 {
-    _address.dword = address;
+    uint32_t& addressRef = reinterpret_cast<uint32_t&>(_address.front());
+	addressRef = address;
 }
 
-IPAddress::IPAddress(const uint8_t *address)
-{
-    memcpy(_address.bytes, address, sizeof(_address.bytes));
-}
+IPAddress::IPAddress(const uint8_t *address) : _address({address[0], address[1], address[2], address[3]})
+{}
 
 IPAddress& IPAddress::operator=(const uint8_t *address)
 {
-    memcpy(_address.bytes, address, sizeof(_address.bytes));
+	std::copy(address, address + _address.size(), _address.begin());
     return *this;
 }
 
 IPAddress& IPAddress::operator=(uint32_t address)
 {
-    _address.dword = address;
+    uint32_t& addressRef = reinterpret_cast<uint32_t&>(_address.front());
```

Comment:
is `_address.front()` guaranteed to be 32-bit aligned?
If not, then this will cause undefined behavior.

Reply:
> is `_address.front()` guaranteed to be 32-bit aligned? If not, then this will cause undefined behavior.

It is guaranteed be aligned as structure -- array is once member in this structure -- it is has same address. It behavior same as union or c-style array (uint8_t a[4]).

Standart says (https://timsong-cpp.github.io/cppwp/n4659/array):

> class template for storing fixed-size sequences of objects. An array is a [contiguous container](https://timsong-cpp.github.io/cppwp/n4659/container.requirements.general#def:contiguous_container). An instance of array<T, N> stores N elements of type T, so that size() == N is an invariant.

> **An array is an [aggregate](https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#def:aggregate) that can be list-initialized with up to N elements whose types are convertible to T.**

https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#1:

> An [aggregate](https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#def:aggregate) is an array or a [class](https://timsong-cpp.github.io/cppwp/n4659/class) with
> 
> [(1.1)](https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#1.1) no user-provided, explicit, or inherited constructors ([[class.ctor]](https://timsong-cpp.github.io/cppwp/n4659/class.ctor)),
> [(1.2)](https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#1.2) no private or protected non-static data members (Clause [[class.access]](https://timsong-cpp.github.io/cppwp/n4659/class.access)),
> [(1.3)](https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#1.3) no [virtual functions](https://timsong-cpp.github.io/cppwp/n4659/class.virtual), and
> [(1.4)](https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#1.4) no virtual, private, or protected base classes ([[class.mi]](https://timsong-cpp.github.io/cppwp/n4659/class.mi)).
> [ Note: Aggregate initialization does not allow accessing protected and private base class' members or constructors.  — end note ]

https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#2:

> 2     The [elements](https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#def:aggregate,elements) of an aggregate are:
> 
> [(2.1)](https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#2.1) for an array, the array elements in increasing subscript order, or
> [(2.2)](https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#2.2) for a class, the direct base classes in declaration order, followed by the direct non-static data members ([[class.mem]](https://timsong-cpp.github.io/cppwp/n4659/class.mem)) that are not members of an anonymous union, in declaration order.

and (https://timsong-cpp.github.io/cppwp/n4659/container.requirements#general-6):

> begin() returns an iterator referring to the first element in the container. end() returns an iterator which is the past-the-end value for the container. If the container is empty, then begin() == end().

https://timsong-cpp.github.io/cppwp/n4659/sequence.reqmts#tab:containers.sequence.optional:

> a.front() = *a.begin()

It mean that `&_address.front() == &(*_address.begin()) == &_address == &IPAddress`
and correspondingly an align of `_address.front()` depend on an align of `IPAddress` structure.

Reply:
but [this example](https://godbolt.org/z/rehMs66We) shows that `alignof` = 1 for struct with `std::array<uint8_t,4>` member.
`alignas(4) `is required to achieve 32-bit alignment
```
struct alignas(4) Sarray { 
std::array<uint8_t,4> a{};
};
```

Reply:
but [this example](https://godbolt.org/z/GGhKe65T6) shows that class `IPAddress` has 4 bytes of aligment because it inherits from an abstract type (with vtable).

in this case, the _address field still has 1 byte of alignment. -- need `alignas( alignof(uint32_t) )` for `IPAddress`.
```
struct IPAddress: public Printable
{
    alignas( alignof(uint32_t) ) std::array<uint8_t,4> _address{};
};
```
[as  here](https://godbolt.org/z/cfncPhefW)

I didn’t initially take into account that inheritance occurs

Reply:
[i fix this omission.](https://github.com/espressif/arduino-esp32/pull/8819/commits/5d65a547906388e2dcb866549c8927d506513a8b)

Reply:
here, of course, I would like to generally highlight that `uint32_t` may not exist at all on a specific platform.
Maybe `uint_least32_t` should be used for compatibility?

Reply:
`uint32_t` is known in all platforms this repository covers.
I don't know `uint_least32_t`, but it seems to be not guaranteed to be a fixed length type, which will cause issues when it is used in some struct of fixed length or offset of members (e.g. storing a struct or transferring it to another node)

Reply:
> I don't know `uint_least32_t`, but it seems to be not guaranteed to be a fixed length type, which will cause issues when it is used in some struct of fixed length or offset of members (e.g. storing a struct or transferring it to another node)

https://timsong-cpp.github.io/cppwp/n4659/cstdint#syn -- types that are at least N-bit